backend/cmm_helpers.ml

```diff
@@ -749,35 +751,46 @@ let mod_int c1 c2 is_safe dbg =
 (* Division or modulo on boxed integers. The overflow case min_int / -1 can
    occur, in which case we force x / -1 = -x and x mod -1 = 0. (PR#5513). *)
 
-let is_different_from x = function
-  | Cconst_int (n, _) -> n <> x
-  | Cconst_natint (n, _) -> n <> Nativeint.of_int x
-  | _ -> false
+(* Division or modulo on boxed integers. The overflow case min_int / -1 can
+   occur, in which case we force x / -1 = -x and x mod -1 = 0. (PR#5513). *)
 
-let safe_divmod_bi mkop kind is_safe mkm1 c1 c2 bi dbg =
-  bind "divisor" c2 (fun c2 ->
-      bind "dividend" c1 (fun c1 ->
-          let c = mkop c1 c2 is_safe dbg in
-          if Arch.division_crashes_on_overflow
-             && bi <> Primitive.Unboxed_int32
-             && not (is_different_from (-1) c2)
-          then
-            Cifthenelse
-              ( Cop (Ccmpi Cne, [c2; Cconst_int (-1, dbg)], dbg),
-                dbg,
-                c,
-                dbg,
-                mkm1 c1 dbg,
-                dbg,
-                kind )
-          else c))
+let safe_divmod_bi mkop mkm1 ?(dividend_cannot_be_min_int = false) is_safe
+    dividend divisor dbg =
+  let is_different_from x = function
+    | Cconst_int (n, _) -> Nativeint.of_int n <> x
+    | Cconst_natint (n, _) -> n <> x
+    | _ -> false
+  in
+  bind "divisor" divisor (fun divisor ->
+      bind "dividend" dividend (fun dividend ->
+          let c = mkop dividend divisor is_safe dbg in
+          if not Arch.division_crashes_on_overflow
+          then c
+          else
+            let dividend_cannot_be_min_int =
+              dividend_cannot_be_min_int
+              || is_different_from Nativeint.min_int dividend
+            in
+            let divisor_cannot_be_negative_one =
+              is_different_from (-1n) divisor
+            in
+            if dividend_cannot_be_min_int || divisor_cannot_be_negative_one
+            then c
+            else
+              Cifthenelse
+                ( Cop (Ccmpi Cne, [divisor; Cconst_int (-1, dbg)], dbg),
+                  dbg,
+                  c,
+                  dbg,
+                  mkm1 dividend dbg,
+                  dbg,
+                  Any )))
 
-let safe_div_bi is_safe =
-  safe_divmod_bi div_int Any is_safe (fun c1 dbg ->
+let safe_div_bi =
```

Comment:
I wouldn't mind an eta-expension here for clarity (and it might also be generate slightly better code though that's not a real concern here I'd say).